### PR TITLE
Implementation MOD-IR-TEMP

### PR DESCRIPTION
**Pull request by fel115**
Issue: #62 

@markus2330 please review, the functionality. What should be implemented in handle()?  When should something be triggered? There are two temperature values the ambient and the temperature of an object, how should they be used?

### CODE REVIEW

**Comment by markus2330** [comment]
> What should be implemented in handle()?

In the current module system it always should update the values (to be stored in the struct)

> When should something be triggered?

When the value changes according the use cases.

> There are two temperature values the ambient and the temperature of an object, how should they be used?

If one of them goes below the threshold it should trigger and only if both go below the warning should be cancelled.

**Comment by fel115** [comment]
Need to be tested, waiting for Nextcloud to be reachable.
# TODO if code works
- [ ] implement Translation

**Comment by markus2330** [comment]
Nextcloud (and everything else) is up again.

File: lang/de_AT.po [hunk 1]
```diff
@@ -19,184 +19,224 @@ msgstr ""
...
+msgid "⚠️ Failed to init MOD-IR-TEMP: {}"
+msgstr "⚠️ Initialisierung von MOD-IR-TEMP ist fehlgeschlagen: {}"
+
+#: src/main.rs:233
+msgid "⚠️ Failed to init MOD-IR-TEMP: ChecksumMismatch"
+msgstr "⚠️ Initialisierung von MOD-IR-TEMP ist fehlgeschlagen: ChecksumMismatch"
+
+#: src/main.rs:236
+msgid "⚠️ Failed to init MOD-IR-TEMP: InvalidInputData"
+msgstr "⚠️ Initialisierung von MOD-IR-TEMP ist fehlgeschlagen: InvalidInputData"
```
Comment (markus2330) [inline lang/de_AT.po:77]: Shouldn't be needed? (Generic one enough)

File: src/main.rs [hunk 2]
```diff
@@ -212,6 +215,29 @@ fn main() -> Result<(), Error> {
 	if config.get_option::<String>("sensors/#0/loc").is_some() {
 		let mut environment = Environment::new(&mut config);
 
+		let mut ir_temp = ModIrTemps::new_default();
```
Comment (markus2330) [inline src/main.rs:218]: Good placement but the constructor's name is different. Please also be consistent in singular name of modules. I think ModIR is enough. Pass config to constructor as for other modules.

```suggestion
		let mut ir_temp = ModIR::new(&mut config);
```

File: src/main.rs [hunk 3]
```diff
@@ -212,6 +215,29 @@ fn main() -> Result<(), Error> {
...
+		) {
+			Ok(modirtemps_obj) => {
+				ir_temp = modirtemps_obj;
+			}
+			Err(error_typ) => match error_typ {
+				MlxError::I2C(error) => {
+					nc.ping(gettext!("⚠️ Failed to init MOD-IR-TEMP: {}", error));
+				}
+				MlxError::ChecksumMismatch => {
+					nc.ping(gettext("⚠️ Failed to init MOD-IR-TEMP: ChecksumMismatch"));
```
Comment (markus2330) [inline src/main.rs:233]: Maybe this already fixes the problem of multiple translation strings?

```suggestion
					nc.ping(gettext("⚠️ Failed to init MOD-IR: {}", "ChecksumMismatch"));
```

File: src/mod_ir_temp.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+	ChangedToCancelled,
+}
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum IrTempState {
+	Normal,
+	TooHot,
+}
+
+pub struct ModIrTemps {
```
Comment (markus2330) [inline src/mod_ir_temp.rs:25]: Please add some documentation to the individual elements.

```suggestion
pub struct ModIR {
```

File: src/mod_ir_temp.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+	mlx: Option<Mlx9061x<I2cdev, Mlx90614>>,
+	device: String,
+	addr: SlaveAddr,
+	pub ambient_temp: f32,
+	pub object_temp: f32,
+	_emissivity: f32,
```
Comment (markus2330) [inline src/mod_ir_temp.rs:31]: Why underscore?

File: src/mod_ir_temp.rs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	active_ambient_state: IrTempState,
+	active_object_state: IrTempState,
+	data_interval: u16,
+	read_counter: u16,
+}
+
+impl ModIrTemps {
+	//Generate default instanz of struct.
```
Comment (markus2330) [inline src/mod_ir_temp.rs:39]:
```suggestion
	//Generate default instance of struct.
```

File: src/mod_ir_temp.rs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+					return Ok(self.set_handle_output(ambient_state, object_state));
+				}
+			}
+			None => (),
+		}
+		Ok(IrTempStateChange::None)
+	}
+
+	//This function sets the emissivity for the measurement of the object temperture
+	//The parameter emissivity can be choosen between 0.0 and 1.0
```
Comment (markus2330) [inline src/mod_ir_temp.rs:173]: How did you choose it and why?

File: src/main.rs [hunk 8]
```diff
@@ -212,6 +215,29 @@ fn main() -> Result<(), Error> {
 	if config.get_option::<String>("sensors/#0/loc").is_some() {
 		let mut environment = Environment::new(&mut config);
 
+		let mut ir_temp = ModIrTemps::new_default();
+
+		match ModIrTemps::new(
+			config.get::<String>("ir_temp/device"),
```
Comment (markus2330) [inline src/main.rs:221]: spec missing. And I would call it mod/ir or simply ir.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thx for the the module, mostly good work, nice user-facing strings. Probably it is best if you first decide about the interface of how modules should look like (constructor, handle) so that we soon get consistency.

General documentation about the mod-ir module (.md file) and spec file is missing.

File: src/mod_ir_temp.rs (hunk 7 above)
Comment (fel115) [reply src/mod_ir_temp.rs:173]: I used the default value, which is set to `1`. However, the emissivity only needs to be adjusted if we are using a specific object for measurement, as indicated [here](https://en.wikipedia.org/wiki/Emissivity). Therefore, I incorporated this function to allow for setting the emissivity when required in certain use cases.

File: src/mod_ir_temp.rs (hunk 5 above)
Comment (fel115) [reply src/mod_ir_temp.rs:31]: To remove the warning of emissivity isn't used.

File: lang/de_AT.po (hunk 1 above)
Comment (fel115) [reply lang/de_AT.po:77]: The translation is not needed for these two?

Comment (fel115) [reply lang/de_AT.po:77]: Ok, I think I know what you mean.

File: src/mod_ir_temp.rs (hunk 7 above)
Comment (markus2330) [reply src/mod_ir_temp.rs]: Great, please add this info to this comment. It is okay for now if it stays hard-coded to 1.0 for now.

File: src/mod_ir_temp.rs [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+			Err(error) => {
+				return Err(error);
+			}
+		}
+		Ok(())
+	}
+
+	/// This function checks if a state change has occurred and returns the corresponding state change.
+	/// It returns 'None' if nothing has changed since the last duration.
+	/// It returns 'ChangedToBothToHot' if both thresholds have been exceeded.
```
Comment (markus2330) [inline src/mod_ir_temp.rs:110]: Some o missing (also in other lines):

```suggestion
	/// It returns 'ChangedToBothTooHot' if both thresholds have been exceeded.
```

**Review by markus2330 — APPROVED** [review]
Let us get this merged before the refactoring. Please rebase/fix conflicts.

File: files/opensesame.spec [hunk 10]
```diff
@@ -173,3 +173,17 @@ description = Which Opensensemap senseBoxes should be connected to the weather s
...
+example=1
+default=0
+
+[ir/device]
+description=Which device to use for the MOD-IR-TEMP sensor. /dev/null means that no MOD-IR-TEMP sensor is connected.
+example = "/dev/i2c-2"
+default = "/dev/null"
+
+[ir/data/interval]
+description=How often to get new data, depends on how often handle() is called
```
Comment (markus2330) [inline files/opensesame.spec:187]: This should be defined by the implementation and documented here.

File: files/opensesame.spec [hunk 11]
```diff
@@ -173,3 +173,17 @@ description = Which Opensensemap senseBoxes should be connected to the weather s
...
+default=60
```
Comment (markus2330) [inline files/opensesame.spec:188]: If this is simply a number in seconds or similar it would be easier for the user.

**Review by markus2330 — APPROVED** [review]
Small problem in the spec but this can be also fixed while refactoring.

**Comment by markus2330** [comment]
Great work, looking forward to progress in refactoring soon! :rocket:

File: files/opensesame.spec (hunk 11 above)
Comment (fel115) [reply files/opensesame.spec:188]: For now, it depends on how often the handle function is called. This will be different in the async approach.

**Pull request merged**